### PR TITLE
Draft: React 18 support

### LINKED ISSUE #337: Error with react 18
### Description

seems related: https://github.com/inlet/react-pixi/issues/306

```
react-dom.development.js:22665 Uncaught Error: Expected subtree parent to be a mounted class component. This error is likely caused by a bug in React. Please file an issue.
    at findCurrentUnmaskedContext (react-pixi.es-dev.js:3904:15)
    at getContextForSubtree (react-pixi.es-dev.js:20484:23)
    at Object.updateContainer (react-pixi.es-dev.js:20603:17)
    at Stage.componentDidMount (react-pixi.es-dev.js:22880:17)
```

### Steps to reproduce

1. Use the code from https://reactpixi.org/components/graphics
2. The error above appears.

### Additional info

- `@inlet/react-pixi` version: 6.7.0
- `React` version: 18.0.0
- `ReactDOM` version: 18.0.0
- `PIXI` version: 6.3.0

### PR DESCRIPTION
**Pull request by saitonakamura**
Hello @inlet , thanks for your amazing work! I've started working on updating react-pixi to support react 18. We're currently using react-pixi along with react-three-fiber (for webxr stuff) and we're experiencing performance problems. R3F (react-three-fiber) is riding the concurrent train for a while so we decided to give it a shot, but react-pixi (among other things though) is blocking us right now. It's only a draft but i got the first version working. The main painpoint is that docz is pretty much dead and i'm not sure picking it up to support react 18 worth the time. I'm currently looking into using storybook with mdx addon, so let's see how it goes. I'll keep you posted and feel free to comment

Fixes #337 

- [x] Update `src/render/index.js` to react 18
- [x] Update `src/stage/index.js` to react 18
- [x] Update all stories to storybook
- [x] Fix prod storybook build
- [ ] Fix netlify preview
- [ ] Update docs to use new api
- [ ] Update codepens
- [x] Fix tests
- [ ] Fix Suspense tests
- [ ] Fix 'receives different events in different containers' test

### CODE REVIEW

**Comment by inlet** [comment]
Hi @saitonakamura,

Thanks this is amazing! Yeah Docz is a pain, I'm happy to ditch it for something else.
Maybe switch to [docusaurus](https://docusaurus.io/)? 

Thanks for collaborating on this. I'm currently pretty stacked with work for the coming two week.. when I got time I can absolutely help you with this PR.

Thanks and we'll be in touch
Patrick

**Comment by spassvogel** [comment]
@saitonakamura is there any update on this?

**Comment by saitonakamura** [comment]
@spassvogel if you want to try this out, i have a package `npm i @saitonakamura/react-pixi@7.0.0-alpha`

If you use TS you can use paths feature, otherwise use webpack aliases or something like that to redirect `@inlet/react-pixi` to `@saitonakamura/react-pixi`

I do this in `tsconfig.json`
```
"paths": {
  "@inlet/react-pixi": ["../node_modules/@saitonakamura/react-pixi"],
}
```

The api has changed to
```
import { createRoot, Container } from '@inlet/react-pixi`

const canvas = document.createElement('canvas')
document.appendChild(canvas)

const root = createRoot(canvas, { width: 800, height: 600 })
root.render(<Container></Container>)
```
Now app is created under the hood (it's a subject to change, I'm not sure about it)

**Comment by spassvogel** [comment]
Hmm, I tried, but my project uses Create React App, and it gives the following error and removes the 'paths' section from tsconfig :(

>The following changes are being made to your tsconfig.json file:
>  - compilerOptions.paths must not be set (aliased imports are not supported)

**Comment by Heilemann** [comment]
Any chance this is coming to the official release soon? Beginning to run into dependency hell 😄

**Comment by MonsieurNaexec** [comment]
Hi,

I'm having issues while trying to execute `npm i @saitonakamura/react-pixi@7.0.0-alpha`
The full output is the following:
```
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: @saitonakamura/react-pixi@7.0.0-alpha
npm WARN Found: prop-types@15.8.1
npm WARN node_modules/prop-types
npm WARN   prop-types@"^15.8.1" from eslint-plugin-react@7.30.0
npm WARN   node_modules/eslint-plugin-react
npm WARN     eslint-plugin-react@"^7.27.1" from eslint-config-react-app@7.0.1
npm WARN     node_modules/eslint-config-react-app
npm WARN
npm WARN Could not resolve dependency:
npm WARN peer prop-types@"^18.0.0" from @saitonakamura/react-pixi@7.0.0-alpha
npm WARN node_modules/@saitonakamura/react-pixi
npm WARN   @saitonakamura/react-pixi@"*" from the root project
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: @saitonakamura/react-pixi@7.0.0-alpha
npm WARN Found: prop-types@undefined
npm WARN node_modules/prop-types
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer prop-types@"^18.0.0" from @saitonakamura/react-pixi@7.0.0-alpha
npm WARN node_modules/@saitonakamura/react-pixi
npm WARN   @saitonakamura/react-pixi@"*" from the root project
npm ERR! code ETARGET
npm ERR! notarget No matching version found for prop-types@^18.0.0.
npm ERR! notarget In most cases you or one of your dependencies are requesting
npm ERR! notarget a package version that doesn't exist.
```
In fact, when I inspect the package archive, the `package.json` contains this:
```json
  "peerDependencies": {
    "pixi.js": "^6.0.0",
    "prop-types": "^18.0.0",
    "react": "^18.0.0",
    "react-dom": "^18.0.1"
  },
```
But the `prop-types` package doesn't have a 18.0.0 version.
I'm not very familiar with package dependencies so maybe there's something I get wrong, but I wonder why nobody reported this issue. Is it related to the Windows environment? Or did I miss something during the installation process?
Also, I'm using CRA with the typescript template.

Thank you for any hint

**Comment by pluveto** [comment]
I'm using @saitonakamura/react-pixi. it works well. waiting for a official release

**Comment by dmytro-podolianskyi-ew** [comment]
> Hi,
> 
> I'm having issues while trying to execute `npm i @saitonakamura/react-pixi@7.0.0-alpha` The full output is the following:
> 
> ```
> npm WARN ERESOLVE overriding peer dependency
> npm WARN While resolving: @saitonakamura/react-pixi@7.0.0-alpha
> npm WARN Found: prop-types@15.8.1
> npm WARN node_modules/prop-types
> npm WARN   prop-types@"^15.8.1" from eslint-plugin-react@7.30.0
> npm WARN   node_modules/eslint-plugin-react
> npm WARN     eslint-plugin-react@"^7.27.1" from eslint-config-react-app@7.0.1
> npm WARN     node_modules/eslint-config-react-app
> npm WARN
> npm WARN Could not resolve dependency:
> npm WARN peer prop-types@"^18.0.0" from @saitonakamura/react-pixi@7.0.0-alpha
> npm WARN node_modules/@saitonakamura/react-pixi
> npm WARN   @saitonakamura/react-pixi@"*" from the root project
> npm WARN ERESOLVE overriding peer dependency
> npm WARN While resolving: @saitonakamura/react-pixi@7.0.0-alpha
> npm WARN Found: prop-types@undefined
> npm WARN node_modules/prop-types
> npm WARN 
> npm WARN Could not resolve dependency:
> npm WARN peer prop-types@"^18.0.0" from @saitonakamura/react-pixi@7.0.0-alpha
> npm WARN node_modules/@saitonakamura/react-pixi
> npm WARN   @saitonakamura/react-pixi@"*" from the root project
> npm ERR! code ETARGET
> npm ERR! notarget No matching version found for prop-types@^18.0.0.
> npm ERR! notarget In most cases you or one of your dependencies are requesting
> npm ERR! notarget a package version that doesn't exist.
> ```
> 
> In fact, when I inspect the package archive, the `package.json` contains this:
> 
> ```json
>   "peerDependencies": {
>     "pixi.js": "^6.0.0",
>     "prop-types": "^18.0.0",
>     "react": "^18.0.0",
>     "react-dom": "^18.0.1"
>   },
> ```
> 
> But the `prop-types` package doesn't have a 18.0.0 version. I'm not very familiar with package dependencies so maybe there's something I get wrong, but I wonder why nobody reported this issue. Is it related to the Windows environment? Or did I miss something during the installation process? Also, I'm using CRA with the typescript template.
> 
> Thank you for any hint

I have the same issue.

@inlet @saitonakamura Did you have a chance to finish your job with upgrading React to 18?

Or @saitonakamura Could you change  "prop-types": "^18.0.0", to the correct version 15.8.1?

**Comment by pluveto** [comment]
@dmytro-podolianskyi-ew I don't have prop-types dependency and looks like it works. you can try to remove that entry

**Comment by MonsieurNaexec** [comment]
> I'm having issues while trying to execute `npm i @saitonakamura/react-pixi@7.0.0-alpha`

I tried using yarn instead of npm, and it seems to work now.
Having to change package managers is quite annoying though...

**Comment by saitonakamura** [comment]
Sure will do next week. Sorry for not devoting time to it, got mixed up with other stuff

**Comment by Bevebage** [comment]
Hey, is there an eta on the release? I just want to know if I should wait a day or two for this or just install now and upgrade later. Thanks so much.

**Comment by inlet** [comment]
We’re working on this in our spare time 😁 no eta yet for the release so yes install and upgrade when it’s available is probably a good idea!

**Comment by wallynm** [comment]
Hey @saitonakamura, i would like to help here, I just need some context to understand how to help. 
Maybe a markdown file, with listed tasks we need to addres?

**Comment by hevans90** [comment]
Thank you so much for working on this, it will be a game-changer for my project!

**Comment by wallynm** [comment]
Hey @saitonakamura , just sent a PR to your repo, https://github.com/wisebits-tech/react-pixi/pull/1
That's not much yet, but im gonna open others pr's along the week, if you see any issues just comment in the PR plx

**Comment by inlet** [comment]
Great to see you guys are collab on this, let me know when I can review a PR

**Comment by saitonakamura** [comment]
I've published `@saitonakamura/react-pixi@7.0.0-alpha-2`, the `createRoot` api (which were [changed initially](https://github.com/inlet/react-pixi/pull/338#issuecomment-1119589794)) returned back to previous design with `app.stage` as a root container

```
import { createRoot, Text } from '@inlet/react-pixi';
import { Application } from 'pixi.js';

const app = new Application({
  width: 800,
  height: 600,
  backgroundColor: 0x10bb99,
  view: document.getElementById('my-canvas'),
});

const root = createRoot(app.stage);
root.render(<Text text="Hello World" x={200} y={200} />);
```

You check out the basic example with `createRoot` at https://codesandbox.io/s/j112mb?file=/src/index.jsx

`<Stage>` component should work as before

cc @spassvogel @Bevebage @hevans90 @pluveto 

I also update `prop-types` version to a proper one, cc @pluveto @MonsieurNaexec

**Comment by inlet** [comment]
@saitonakamura is it stable enough to have this PR merged? Thanks!

**Comment by saitonakamura** [comment]
@inlet Nah, a lot of tests still failing. But if you can release to npm under next tag or something like that it would certainly be great for the folk out there. Because it's stable enough so the examples and stories works

**Comment by wallynm** [comment]
@saitonakamura, next PR i gonna try to start fixing tests so we can ensure a better version!
I think today i can make some PR's improving tests

**Comment by saitonakamura** [comment]
We're slowly getting there guys

**Comment by Specy** [comment]
Wait will this new version still with for older react versions?

**Comment by saitonakamura** [comment]
@Specy no, this one only for react 18
But this consists of nothing new but support for react 18. If you use older versions you can just use react-pixi@6.*

Eventually we will stumble upon problems when we'll do fixes or new features/ cause we'll need to backport them to the other version . But let the future us worry about it

**Comment by saitonakamura** [comment]
Only 4 more test suits to go!

**Comment by mutewinter** [comment]
I'm using `@saitonakamura/react-pixi@7.0.0-alpha-4` and I spotted an edge-case bug:

On iOS Safari (tested on iOS 15.6 and iPadOS 15.6) React 18's new double effect firing in `[StrictMode](https://reactjs.org/docs/strict-mode.html)` causes the Application to be created twice in quick succession. This causes the GL context Pixi uses to be created twice, as well. In turn, this causes the Pixi app to throw with the error: `Invalid value of `0` passed to `checkMaxIfStatementsInShader`.

I temporarily fixed this by disabling strict mode. However, it seems like one would want to reuse Pixi Applications across subsequent renders like this.

To reproduce: run this branch with React 18 and [StrictMode](https://reactjs.org/docs/strict-mode.html] enabled.

**Comment by saitonakamura** [comment]
@mutewinter are you running your app with dev react build and strict mode? Also, do you use `<Stage>` or `createRoot` api?

**Comment by saitonakamura** [comment]
@mutewinter can you do a codesandbox of something like that illustrating that issue?

**Comment by saitonakamura** [comment]
Hey folks, providing an update on this.
I fixed the majority of the tests, only three tests were skipped
image
Good news are, 99% of the tests fixes were done in the tests bodies themselves
So I released the `@saitonakamura/react-pixi@7.0.0-beta-1`
I'll continue migrating with the stories and updating the docs in the meantime

@inlet if you have time, I'd like to take a look at the skipped tests (marked in the starting message and provided with TODO explanations in the code), maybe you'll have some ideas

**Comment by mutewinter** [comment]
> @mutewinter are you running your app with dev react build and strict mode? Also, do you use `<Stage>` or `createRoot` api?

Dev React build via Next.js which is using `createRoot`.

**Comment by mutewinter** [comment]
> @mutewinter can you do a codesandbox of something like that illustrating that issue?

Here's the sandbox: https://codesandbox.io/s/react-pixi-18-and-strictmode-fails-on-ios-fg44ky?file=/src/index.jsx

Visit <https://fg44ky.csb.app/> on iOS and you'll get the error: `Invalid value of 0 passed to checkMaxIfStatementsInShader`

**Comment by saitonakamura** [comment]
@mutewinter , I found out that it's reproducible by simply creating and destroying 2 pixi apps using the same canvas, see https://codesandbox.io/s/gracious-water-nmiswl?file=/src/index.js, https://nmiswl.csb.app/. Maybe someone knows how to workaround it?

**Comment by joukosaastamoinen** [comment]
Hey! Thanks for the awesome work here. I'm trying to use `@saitonakamura/react-pixi@7.0.0-beta-1` in a TypeScript project and I get this error:

```
Module '"@saitonakamura/react-pixi"' has no exported member 'createRoot'.
```

Is the type declarations incorrect/out-of-date?

**Comment by saitonakamura** [comment]
@joukosaastamoinen yeah, I've yet to update the d.ts, thanks for the report

**Comment by mutewinter** [comment]
https://github.com/wisebits-tech/react-pixi/blob/284e34390e744457cb0e6fd49c2be210b6928b76/src/render/index.jsx#L68

This warning appears to have the conditional inverted.

**Comment by saitonakamura** [comment]
Released `@saitonakamura/react-pixi@7.0.0-beta-2`. Only 2 fixes were made in the lib code: d.ts fix and fix of wrong warning

Other commits were related to tests, stories and stories deployment

cc @mutewinter @joukosaastamoinen

**Comment by saitonakamura** [comment]
@inlet can you help me with the netlify? deploy looks to be successful, but when storybook is trying to load the assets it fails with 404. I check the storybook build locally and it's fine
image

File: index.d.ts [hunk 1]
```diff
@@ -342,49 +329,58 @@ declare namespace _ReactPixi {
...
-export const BitmapText: React.FC<_ReactPixi.IBitmapText>;
-export const NineSlicePlane: React.FC<_ReactPixi.INineSlicePlane>;
-export const ParticleContainer: React.FC<_ReactPixi.IParticleContainer>;
-export const TilingSprite: React.FC<_ReactPixi.ITilingSprite>;
-export const SimpleRope: React.FC<_ReactPixi.ISimpleRope>;
-export const SimpleMesh: React.FC<_ReactPixi.ISimpleMesh>;
-export const AnimatedSprite: React.FC<_ReactPixi.IAnimatedSprite>;
+export const Text: React.FC<_ReactPixi.IText>
+export const Sprite: React.FC<_ReactPixi.ISprite>
+export const Container: React.FC<_ReactPixi.IContainer>
```
Comment (ryohey) [inline index.d.ts:346]:
```suggestion
export const Container: React.FC<React.PropsWithChildren<_ReactPixi.IContainer>>
```
Since React18, FC no longer includes children, Maybe we need PropsWithChildren? other components as well.

Comment (saitonakamura) [reply index.d.ts:346]: @ryohey yeah, you're right, thanks for the suggestion!

Comment (saitonakamura) [reply index.d.ts]: Fixed

**Comment by mutewinter** [comment]
A new small issue with this branch vs main: `window` is accessed here: https://github.com/wisebits-tech/react-pixi/blob/39e23a3dfa09ff032723815ca73a304708423d90/src/reconciler/hostconfig.js#L101, which causes rendering in Node to fail without a polyfill. Previously, one could get away without a polyfill of window by providing `resolution` and `autoDensity: false`. This code is [commented in such a way](https://github.com/wisebits-tech/react-pixi/blob/39e23a3dfa09ff032723815ca73a304708423d90/src/reconciler/hostconfig.js#L348) that it doesn't appear done, so this may be on the todo list.

**Comment by saitonakamura** [comment]
@mutewinter awesome that you've found it, I love how y'all helping with the testing!
Implementation of `getEventPriority` is just copied from the react-dom counterpart, TODO is left to think if we can do better. But you're indeed right that it should not fail the node so I'm gonna insert a check beforehand for now

**Comment by inlet** [comment]
> @inlet can you help me with the netlify? deploy looks to be successful, but when storybook is trying to load the assets it fails with 404. I check the storybook build locally and it's fine
> image

I’m currently on holiday but will help when I’m back, thanks 🙏

**Comment by mutewinter** [comment]
> @mutewinter awesome that you've found it, I love how y'all helping with the testing!
> Implementation of `getEventPriority` is just copied from the react-dom counterpart, TODO is left to think if we can do better. But you're indeed right that it should not fail the node so I'm gonna insert a check beforehand for now

Here's a patch that fixes the crash when rendering in Node:

```patch
index 7cbff42..700f3b7 100644
--- a/src/reconciler/hostconfig.js
+++ b/src/reconciler/hostconfig.js
@@ -98,6 +98,10 @@ function diffProperties(pixiElement, type, lastProps, nextProps, rootContainerEl
 }
 
 export function getEventPriority() {
+  // Escape hatch for Node renderer where we don't a Window.
+  if (typeof window === 'undefined') {
+    return DefaultEventPriority
+  }
   let name = window?.event?.type
   switch (name) {
     case 'click':
```

**Comment by saitonakamura** [comment]
@mutewinter published a 5th with your fix

**Comment by nigel-loops** [comment]
Hi guys. My project fundamentally relies on react-pixi. Looking to update to react 18 but see this is still WIP.

What is still required in order to get this over the line for release?

Thanks

**Comment by inlet** [comment]
Once finished I'll be glad to merge the PR!

**Comment by mutewinter** [comment]
Been using this branch in production for weeks and it has been solid for me.

**Comment by alvinalexander** [comment]
I would like to use this package for an internal project at work, but I need support for React 18. What are the remaining todos for this PR?

**Comment by hevans90** [comment]
@alvinalexander you can try using @saitonakamura 's fork that has experimental (but pretty stable) react 18 support here:

https://www.npmjs.com/package/@saitonakamura/react-pixi?activeTab=versions

**Comment by baseten** [comment]
@saitonakamura thanks for your awesome work on this! I'm going to be getting an official release working with React 18 as soon as possible.

We're still deciding how to handle docs and examples along with a general revamp across the whole PIXI ecosystem - it will likely be a combination of docusaurus and storybook. In the short term getting React 18 support working is the top priority and it appears that is generally a small part of this PR (mainly related to `createRoot` changes) compared with docs changes?

**Comment by baseten** [comment]
I've opened a minimal PR just to get the library working with React 18 before we do a docs overhaul in the near future. 

https://github.com/pixijs/pixi-react/pull/358

@saitonakamura I think you've been doing a bigger overhaul including docs that may have extended beyond this original draft PR - it might be useful to sync to discuss?

**Comment by baseten** [comment]
React 18 is now supported, including a matching `createRoot` API as of `v7.0.0` available at `@pixi/react`
